### ISSUE TITLE
Using ES2018 object spread in module.exports results in an internal compiler error

### ISSUE THREAD

**Issue by nichtverstehen** [google/closure-compiler#2926]
Compiling the following code:
```
var g = {};
module.exports = {
  ...g,
};
```
with `--process_common_js_modules` gives the following error message:
```
INTERNAL COMPILER ERROR.

SPREAD 8 [length: 4] [source_file: experimental/users/cyriln/js/a.js] is not a string node
  Node(SCRIPT): experimental/users/cyriln/js/a.js:6:0
var g = {};
  Parent(ROOT): [source unknown]

  Node(SCRIPT): experimental/users/cyriln/js/a.js:6:0
var g = {};
  Parent(ROOT): [source unknown]

	at com.google.javascript.rhino.Node.getString(Node.java:1137)
	at com.google.javascript.jscomp.ProcessCommonJSModules$RewriteModule.expandObjectLitAssignment(ProcessCommonJSModules.java:1452)
	at com.google.javascript.jscomp.ProcessCommonJSModules$RewriteModule.visitExport(ProcessCommonJSModules.java:1332)
	at com.google.javascript.jscomp.ProcessCommonJSModules$RewriteModule.visit(ProcessCommonJSModules.java:1158)
	at com.google.javascript.jscomp.NodeTraversal.handleScript(NodeTraversal.java:730)
        ...
```

This happens because the object spread is not transpiled by the the CommonJS module processor kicks in, and the latter doesn't know how to handle the spread.

**Comment by ChadKillingsworth** [comment]
CommonJS module rewriting is rather involved. The primary goal originally was to avoid creating aliases with the exports as the compiler lost type information on aliases. So for instance:

```js
let foo = 'bar';
let baz = class {};
module.exports = {
  foo,
  baz,
  a: 'a'
};
```

Is rewritten to:

```js
module$test$default.foo = 'bar';
module$test$default.baz = class{};
module$test$default.a = 'a';
```

The object literal expansion isn't an optimization. It's designed to allow specific properties to be rewritten / removed from the export and the declared names be written to the exported name directly.

If the type checker can be improved to the point that these aliases are understood - the pass could be made considerably simpler.

**Comment by nichtverstehen** [comment]
@ChadKillingsworth Won't same considerations apply for any other object initialized through an object literal and aliasing other variables, not just `module$test$default`: properties should be able to be dropped, types should be preserved?

And indeed it appears to work for simple objects, implemented in part by `CollapseProperties` and by other passes.

**Comment by ChadKillingsworth** [comment]
Sure - but developers typically have control over other objects. Module rewriting creates aliases that developers cannot control or fix.

We've improved the type checking story considerably since the major overhaul of CommonJS rewriting. `CollapseProperties` now even runs in simple mode for module exports only.

The other significant challenge is per-file compilation. `require('./test').foo` and `module.exports = {foo: 'bar'}` have to be rewritten identically even if they are part of separate compilations.

**Comment by nichtverstehen** [comment]
@ChadKillingsworth How do you think a fix for this issue could look like?

**Comment by ChadKillingsworth** [comment]
In `expandObjectLitAssignment`, I would just check for a spread operator and skip that property expansion. I think an alias in this case is appropriate.

**Comment by nichtverstehen** [comment]
From what I see, simply skipping won't work, because the whole `module.exports` assignment is completely removed in the end of `expandObjectLitAssignment`.

One option will be to remove members from the object literal as we go converting them to top-level assignments (and then not remove the object literal assignment).

So in my example it would result in
```
module.exports = {...b};
module.exports.a = "foo";
module.exports.c = "bar";
```

That's not completely correct, because `b` might contain member `a` that should have overridden `"foo"`. Probably it's a good start.

Semantically correct output would be:
```
module.exports.a = "foo";
Object.assign(module.exports, b);
module.exports.c = "bar";
```
(but I don't know if it would confuse type-checker).

What do you think, @ChadKillingsworth?

**Comment by ChadKillingsworth** [comment]
If Object.assign works - I would just use that.

**Comment by nichtverstehen** [comment]
@ChadKillingsworth Thank you for the guidance!

To be clear: I'll probably won't be able to work on this in near future. This issue didn't give in to my attempts to quick fix it. My workaround for now is to compile spreads away with `babel-plugin-transform-object-rest-spread`.

So if anyone else can pick this up, please do.

**Comment by ChadKillingsworth** [comment]
I have a list of changes to work on in that pass. I'll see what I can do.

**Comment by ChadKillingsworth** [comment]
Closed by f90603a
